### PR TITLE
Tidy up `CompetitionEstimator` code

### PR DESCRIPTION
**Pull request by MartinquaXD**
# Description
The current `CompetitionEstimator` code is pretty complex and hard to read. Mostly for 2 reasons:
1. generic implementation tries to do everything which is overly complex
2. specalized implementations live in the same file (native, regular)

# Changes
There are various changes in this PR but the most important one is making the generic implementation less generic.
Instead of supporting all requirements of the native and regular price estimator implementations the generic version now only handles producing results, some logging and metrics.
This allows the generic and specialized implementations to be easier to implement and therefore read IMO.

Additional changes:
* we no longer have `RacingCompetitionEstimator` and `CompetitionEstimator`; instead you can turn a `CompetitionEstimator` into a racing one by calling `.with_early_return()` on it
* separated logic into 3 files:
    1. main logic + tests of staging logic (i.e. generic aspect of the code)
    2. code + tests specific to native price estimation
    3. code + tests specific to regular price estimation (also contains bang for buck implementation)
* deleted some unused code

## How to test
All existing tests continue to pass without any changes (besides having to call `.with_early_return()`)

## How to review
It's probably easiest to review the individual commits in order.
Pay extra attention to `produce_results()`. I changed that code the most and the original code (`estimate_generic()`) was not trivial.

### CODE REVIEW

**Comment by MartinquaXD** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by MartinquaXD** [comment]
recheck

File: crates/shared/src/price_estimation/competition/regular.rs [hunk 1]
```diff
@@ -0,0 +1,318 @@
+use {
```
Comment (fleupold) [inline crates/shared/src/price_estimation/competition/regular.rs:1]: "regular" is quite vague. Can we call this `quote` or something more specific?

**Review by fleupold — APPROVED** [review]
> It's probably easiest to review the individual commits in order.

Isn't this a sign that this PR should be submitted as a stack of 1PR per commit ;-) 

Now my comments are kind of awkwardly placed in inner commits which may be hard to contextualize. Main concern is around the parallelisation on the price estimation context.

File: crates/shared/src/price_estimation/competition.rs [hunk 2]
```diff
@@ -59,7 +59,7 @@ pub struct CompetitionEstimator<T> {
 
 impl<T: Send + Sync + 'static> CompetitionEstimator<T> {
     pub fn new(inner: Vec<PriceEstimationStage<T>>, ranking: PriceRanking) -> Self {
-        assert!(!inner.is_empty());
+        assert_ne!(inner.iter().map(Vec::len).count(), 0);
```
[the remark this answers is not in the document]
Comment (MartinquaXD) [reply crates/shared/src/price_estimation/competition.rs]: True also noticed that later on as well.
The latest code is:
```rust
        assert!(!stages.is_empty());
        assert!(stages.iter().all(|stage| !stage.is_empty()));
 ```

**Review by squadgazzz — APPROVED** [review]

**Pull request merged**